### PR TITLE
Add KB section on reducing planning time for many joins.

### PR DESCRIPTION
**Pull request by stbof**
https://cockroachlabs.atlassian.net/browse/DOC-2407

### CODE REVIEW

**Review by cucaroach — APPROVED** [review]
:LGTM: but I would have an expert like @rytaft or @mgartner also sign off on this.

> Reviewed 1 of 1 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/docs/13421#-:-MzQSNvg0Q0fHLpiGgDC:bcbukux)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @stbof)

File: v22.1/cost-based-optimizer.md [hunk 1]
```diff
@@ -158,6 +158,22 @@ We strongly recommend not setting this value higher than 8 to avoid performance
 
 For more information about the difficulty of selecting an optimal join ordering, see our blog post [An Introduction to Join Ordering](https://www.cockroachlabs.com/blog/join-ordering-pt1/).
 
+### Reduce planning time for plans with many joins
+
+The cost-based optimizer explores multiple join orderings to find the lowest-cost plan. If there are many joins or join subtrees in the query, this can increase the number of execution plans the optimizer explores, and therefore the exploration and planning time. If the plan phase of a query takes a long time (on the order of multiple seconds or minutes) to plan, or the query plan involves many joins, consider the following alternatives to reduce the planning time:
```
Comment (mgartner) [inline v22.1/cost-based-optimizer.md:163]: nit: Maybe "If the plan phase..." should be "If the **planning** phase..."?

File: v22.1/cost-based-optimizer.md [hunk 2]
```diff
@@ -158,6 +158,22 @@ We strongly recommend not setting this value higher than 8 to avoid performance
...
+
+- To limit the size of the subtree that can be reordered, set the `reorder_joins_limit` session variable to a lower value, for example:
+
+    ~~~ sql
+    SET reorder_joins_limit = 2;
+    ~~~
+
+    If the ordering is acceptable, for the shortest planning time, you can set `reorder_joins_limit` to `0`.
```
Comment (mgartner) [inline v22.1/cost-based-optimizer.md:171]: nit: "If the **join** ordering **inherent in the query** is acceptable ..."

nit: If we want to be extra clear we can add a sentence like "This disables exploration of join orderings entirely."

File: v22.1/cost-based-optimizer.md [hunk 3]
```diff
@@ -158,6 +158,22 @@ We strongly recommend not setting this value higher than 8 to avoid performance
 
 For more information about the difficulty of selecting an optimal join ordering, see our blog post [An Introduction to Join Ordering](https://www.cockroachlabs.com/blog/join-ordering-pt1/).
 
+### Reduce planning time for plans with many joins
+
+The cost-based optimizer explores multiple join orderings to find the lowest-cost plan. If there are many joins or join subtrees in the query, this can increase the number of execution plans the optimizer explores, and therefore the exploration and planning time. If the plan phase of a query takes a long time (on the order of multiple seconds or minutes) to plan, or the query plan involves many joins, consider the following alternatives to reduce the planning time:
+
+- To limit the size of the subtree that can be reordered, set the `reorder_joins_limit` session variable to a lower value, for example:
```
Comment (mgartner) [inline v22.1/cost-based-optimizer.md:165]: We should make the downsides clear. By reducing `reorder_joins_limit` we reduce the number of plans explored, so a less efficient plan may be chosen by the optimizer.

**Review by mgartner — APPROVED** [review]
Looks great! Just left some minor suggestions.

File: v22.1/cost-based-optimizer.md (hunk 3 above)
Comment (mgartner) [reply v22.1/cost-based-optimizer.md:165]: Did you see this comment? I think it's important to highlight the downside of decreasing `reorder_joins_limit`, but let me know if you disagree.

**Review by mgartner — COMMENTED** [review]
Should we also mention that the cluster setting `sql.defaults.reorder_joins_limit` changes the default `reorder_joins_limit` session setting, and that `sql.defaults.reorder_joins_limit` is `8` by default?

File: v22.1/cost-based-optimizer.md (hunk 3 above)
Comment (stbof) [reply v22.1/cost-based-optimizer.md:165]: I did see the comment and had added the sentence, but it looks like I forgot to commit it. I just did.

**Review by mgartner — APPROVED** [review]
Very nice! LGTM.

File: v22.1/cost-based-optimizer.md [hunk 4]
```diff
@@ -145,19 +145,39 @@ For a query involving multiple joins, the cost-based optimizer will explore addi
...
+
+    By reducing `reorder_joins_limit` CockroachDB reduces the number of plans explored, so a less efficient plan may be chosen by the optimizer.
+
+    If one query has a slow planning time, you can avoid interfering with other query plans by setting `reorder_joins_limit` to the desired lower value before executing that query and resetting the session variable to the default after executing the query.
 
+- If setting and reseting the session variable is cumbersome or if there are multiple independent joins in the query where some may benefit from join reordering, you can use a [join hint](#join-hints). If the join has a hint specifying the type of join to something other than the default `INNER` (i.e. `INNER LOOKUP`, `MERGE`, `HASH`, etc.), join reordering will be disabled and the plan will respect the join order inherent in the way the query is written. This works at the expression level and doesn't affect the entire query (for instance, if you have a union of two joins they are independent join expressions).
```
[the remark this answers is not in the document]
Comment (rmloveland) [reply v22.1/cost-based-optimizer.md:179]: spelling: I think this should be 'resetting'

**Review by rytaft — APPROVED** [review]
:lgtm: just one nit

> Reviewed 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/docs/13421#-:-MzuZB9v7Z-ncgv8QGsG:b4xgs2i)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner and @stbof)

___
*[v22.1/cost-based-optimizer.md, line 179 at r5](https://reviewable.io/reviews/cockroachdb/docs/13421#-MzuYqmjAmd2nTTYl-ZQ:-MzuYqmkE4YkAswtcz3W:bf37jrp) ([raw file](https://github.com/cockroachdb/docs/blob/eb68c23dd91d28b930e06726f04836ea1c430d9c/v22.1/cost-based-optimizer.md#L179)):*
> ```Markdown
>     If one query has a slow planning time, you can avoid interfering with other query plans by setting `reorder_joins_limit` to the desired lower value before executing that query and resetting the session variable to the default after executing the query.
> 
> - If setting and resetting the session variable is cumbersome or if there are multiple independent joins in the query where some may benefit from join reordering, you can use a [join hint](#join-hints). If the join has a hint specifying the type of join to something other than the default `INNER` (i.e. `INNER LOOKUP`, `MERGE`, `HASH`, etc.), join reordering will be disabled and the plan will respect the join order inherent in the way the query is written. This works at the expression level and doesn't affect the entire query (for instance, if you have a union of two joins they are independent join expressions).
> ```

nit: hints work for all types of joins, not just `INNER`

**Comment by stbof** [comment]
> :lgtm: just one nit
> 
> > Reviewed 1 of 1 files at r5, all commit messages.
> > _[Reviewable](https://reviewable.io/reviews/cockroachdb/docs/13421#-:-MzuZB9v7Z-ncgv8QGsG:b4xgs2i)_ status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner and @stbof)
> 
> _[v22.1/cost-based-optimizer.md, line 179 at r5](https://reviewable.io/reviews/cockroachdb/docs/13421#-MzuYqmjAmd2nTTYl-ZQ:-MzuYqmkE4YkAswtcz3W:bf37jrp) ([raw file](https://github.com/cockroachdb/docs/blob/eb68c23dd91d28b930e06726f04836ea1c430d9c/v22.1/cost-based-optimizer.md#L179)):_
> 
> > ```gfm
> >     If one query has a slow planning time, you can avoid interfering with other query plans by setting `reorder_joins_limit` to the desired lower value before executing that query and resetting the session variable to the default after executing the query.
> > 
> > - If setting and resetting the session variable is cumbersome or if there are multiple independent joins in the query where some may benefit from join reordering, you can use a [join hint](#join-hints). If the join has a hint specifying the type of join to something other than the default `INNER` (i.e. `INNER LOOKUP`, `MERGE`, `HASH`, etc.), join reordering will be disabled and the plan will respect the join order inherent in the way the query is written. This works at the expression level and doesn't affect the entire query (for instance, if you have a union of two joins they are independent join expressions).
> > ```
> 
> nit: hints work for all types of joins, not just `INNER`

@rytaft This says that `INNER` is the default join hint, not that it only works for `INNER`.

**Pull request merged**